### PR TITLE
Move analysis for MIR borrowck

### PR DESCRIPTION
**Pull request by pnkfelix**
This PR adds code for doing MIR-based gathering of the moves in a `fn` and the dataflow to determine where uninitialized locations flow to, analogous to how the same thing is done in `borrowck`.

It also adds a couple attributes to print out graphviz visualizations of the analyzed MIR that includes the dataflow analysis results.

cc @nikomatsakis

### CODE REVIEW

**Comment by nikomatsakis** [comment]
:snowman: (yes, this is a positive emoji, don't ask me why)

**Comment by nikomatsakis** [comment]
This basically looks good to me. Left a bunch of nits and minor comments. I'll take a look at the next round and think more deeply.

**Comment by pnkfelix** [comment]
@nikomatsakis BTW one other thing I only thought while rebasing the code (but did not attempt to implement): instead of the newer approach to statics that I used in f244405, I might instead consider trying to handle abstracting all statics (or all L-values that are built off of a static) into a single abstract entity in the lifting step that is currently in [abs_domain.rs](https://github.com/pnkfelix/rust/blob/borrowck-on-mir-move-analysis/src/librustc_borrowck/borrowck/mir/abs_domain.rs)

But I am not 100% sure that actually makes sense. So I left it at something to think about and tackle later.

(Also, something I *do* need/want to do is to put in actual tests, using the same sort of error-message-based instrumentation that we've used elsewhere to make observations about the compiler's internal view of a piece of code... its possible that upkeep for such tests could be a pain, but we'd only need to keep the tests around until the code is exercised by other things.)

**Comment by nikomatsakis** [comment]
@pnkfelix 

> (Also, something I do need/want to do is to put in actual tests, using the same sort of error-message-based instrumentation that we've used elsewhere to make observations about the compiler's internal view of a piece of code... its possible that upkeep for such tests could be a pain, but we'd only need to keep the tests around until the code is exercised by other things.)

strong :+1:

File: src/librustc_borrowck/borrowck/mir/dataflow.rs [hunk 1]
```diff
@@ -0,0 +1,505 @@
...
+                }
+            }
+            repr::Terminator::Call { ref cleanup, ref destination, func: _, args: _ } => {
+                if let Some(ref unwind) = *cleanup {
+                    self.propagate_bits_into_entry_set_for(in_out, changed, unwind);
+                }
+                if let Some((ref dest_lval, ref dest_bb)) = *destination {
+                    // N.B.: This must be done *last*, after all other
+                    // propagation, as documented in comment above.
+                    on_return(&self.operator, in_out, dest_lval);
```
Comment (nikomatsakis) [inline src/librustc_borrowck/borrowck/mir/dataflow.rs]: Seems...pretty clean.

**Comment by nikomatsakis** [comment]
r=me on the revisions

File: src/librustc_borrowck/borrowck/mir/mod.rs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+            self.mir.basic_block_data(bb);
+        for stmt in statements {
+            self.process_statement(bb, stmt);
+        }
+
+        self.process_terminator(bb, terminator);
+    }
+
+    fn process_statement(&mut self, bb: BasicBlock, stmt: &Statement<'tcx>) {
+        debug!("MirBorrowckCtxt::preprocess_statement({:?}, {:?}", bb, stmt);
```
Comment (dotdash) [inline src/librustc_borrowck/borrowck/mir/mod.rs]: Function name is wrong in the debug message, `preprocess` vs. `process`

File: src/librustc_borrowck/borrowck/mir/mod.rs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    fn process_terminator(&mut self, bb: BasicBlock, term: &Option<Terminator<'tcx>>) {
+        debug!("MirBorrowckCtxt::preprocess_terminator({:?}, {:?})", bb, term);
```
Comment (dotdash) [inline src/librustc_borrowck/borrowck/mir/mod.rs]: Same here

**Comment by pnkfelix** [comment]
Hmm  I assume this is some rustbuild dependency issue; will look

**Pull request merged**